lib/geordi/docker.rb

```diff
@@ -0,0 +1,116 @@
+require 'geordi/interaction'
+require 'geordi/cucumber'
+require 'yaml'
+
+module Geordi
+  class Docker
+    DOCKER_COMPOSE_FILE = 'docker-compose.yml'.freeze
+
+    include Interaction
+
+    def setup
+      check_installation_and_config
+      announce('Building containers...')
+      if execute(:system, 'docker-compose', 'pull')
+        success('Build successful.')
+      else
+        fail('Build failed.')
+      end
+    end
+
+    def shell(options = {})
+      check_installation_and_config
+      if options[:secondary]
+        attach_to_running_shell
+      else
+        run_shell
+      end
+    end
+
+    def vnc
+      Cucumber.new.launch_vnc_viewer('::5967')
+    end
+
+    private
+
+    def attach_to_running_shell
+      running_containers = execute(:`, 'docker-compose ps').split("\n")
+      if (main_container_line = running_containers.grep(/_main_run/).first)
+        container_name = main_container_line.split(' ').first
+        execute(:exec, 'docker', 'exec', '-it', container_name, 'bash')
+      else
+        fail('Could not find a running shell. Start without --secondary first.')
+      end
+    end
+
+    def run_shell
+      command = [:system, 'docker-compose', 'run', '--service-ports']
+      command += ssh_agent_forward
+      command += ['main']
+      execute(*command)
+      execute(:system, 'docker-compose', 'stop')
+    end
+
+    def execute(kind, *args)
+      if ENV['GEORDI_TESTING']
+        puts "Stubbed run #{args.join(' ')}"
+        if kind == :`
+          mock_parse(*args)
+        else
+          mock_run(*args)
+        end
+      else
+        send(kind, *args)
```

Comment:
why don't you always use either `system`, `exec` or ``?

Reply:
Don't quite understand. I need all three variants, since they do different things.
- I want to use exec in case I don't need to do anything after the command returns.
- I use ` if I need the output.
- I use `system` if I want the exitcode.